### PR TITLE
[libc++][type_traits] P2674R1: A trait for implicit lifetime types

### LINKED ISSUE #105259: P2674R1: A trait for implicit lifetime types
**Link:** https://wg21.link/P2674R1

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements P2674R1: https://wg21.link/P2674R1

- https://eel.is/c++draft/type.traits
  - https://eel.is/c++draft/meta.type.synop
  - https://eel.is/c++draft/meta.unary.prop
- https://eel.is/c++draft/support.limits
  - https://eel.is/c++draft/version.syn

Implementation details:
- Uses compiler intrinsic `__builtin_is_implicit_lifetime`:
  - https://github.com/llvm/llvm-project/pull/101807
- Tests based on:
    - https://github.com/llvm/llvm-project/blob/d213981c80626698a07b11ce872acba098a863d4/clang/test/SemaCXX/type-traits.cpp#L1989

References:
- Implicit-lifetime
  - Implicit-lifetime types [basic.types.general]/9: https://eel.is/c++draft/basic.types.general
  - Implicit-lifetime class [class.prop]/9: https://eel.is/c++draft/class.prop
- P0593R6 Implicit creation of objects for low-level object manipulation: https://wg21.link/P0593R6
- P1010R1 Container support for implicit lifetime types: https://wg21.link/P1010R1
- P0593R6 Implicit creation of objects for low-level object manipulation: https://wg21.link/P0593R6

Closes: #105259

### CODE REVIEW

**Comment by MitalAshok** [comment]
The implementation without the builtin can never be implemented correctly for all structs. You wouldn't be able to check if this class has a trivial default constructor:

```c++
struct X {
  X() = default;  // Trivial eligible constructor
  X(int = 0);  // ... But X is not default constructible
  X(const X&);
};

static_assert(__builtin_is_implicit_lifetime(X));
static_assert(!is_implicit_lifetime_v<X>);
```

Not sure you should try to provide a fallback in the `!__has_builtin` case.

(But if you do keep it, you don't check if it's a class before you check `trivially_destructible&&(trivially_X_constructible||...)`. This allows extension types like `double _Complex` and `int [[vector_size(4)]]`, but does not deny references like `int&` which are *not* implicit lifetime)

**Comment by Zingam** [comment]
> The implementation without the builtin can never be implemented correctly for all structs. You wouldn't be able to check if this class has a trivial default constructor:
> 
> ```c++
> struct X {
>   X() = default;  // Trivial eligible constructor
>   X(int = 0);  // ... But X is not default constructible
>   X(const X&);
> };
> 
> static_assert(__builtin_is_implicit_lifetime(X));
> static_assert(!is_implicit_lifetime_v<X>);
> ```
> 
> Not sure you should try to provide a fallback in the `!__has_builtin` case.
> 
> (But if you do keep it, you don't check if it's a class before you check `trivially_destructible&&(trivially_X_constructible||...)`. This allows extension types like `double _Complex` and `int [[vector_size(4)]]`, but does not deny references like `int&` which are _not_ implicit lifetime)

Thanks for the review. I plan to remove the non-builtin implementation.

**Comment by Zingam** [comment]
For the record: I encountered the following issue https://github.com/llvm/llvm-project/issues/106864 while implementing https://github.com/llvm/llvm-project/issues/105259.

**Comment by Zingam** [comment]
Strange build failure with macOS back deployment: https://buildkite.com/llvm-project/libcxx-ci/builds/37470#0191b844-f2b4-4102-b33f-e633e18147c7

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (ldionne) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: This `XFAIL` seems redundant with the `UNSUPPORTED` above.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (ldionne) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:126]: Both of those could be tested with `assert` really. We test the constexpr-friendliness via the top-level `static_assert`.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.verify.cpp
Comment (ldionne) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.verify.cpp:13]: Same comment about `XFAIL`.

**Review by ldionne — COMMENTED** [review]
A few comments but this looks really good!

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: The build fails on Apple back-deployment targets, please check this:
|https://buildkite.com/llvm-project/libcxx-ci/builds/37470#0191b844-f2b4-4102-b33f-e633e18147c7
It looks as if `UNSUPPORTED: apple-clang-16, apple-clang-17` doesn't work here.
I don't know how to solve this. Any ideas?

Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: Would removing the `apple-clang-16, apple-clang-17` from `UNSUPPORTED` be acceptable?

Comment (robincaloudis) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: > The build fails on Apple back-deployment targets, please check this: |https://buildkite.com/llvm-project/libcxx-ci/builds/37470#0191b844-f2b4-4102-b33f-e633e18147c7 It looks as if `UNSUPPORTED: apple-clang-16, apple-clang-17` doesn't work here. Unless I misread the error message. I don't know how to solve this. Any ideas?

+1. I have the exact same question.

Comment (robincaloudis) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: I tried using UNSUPPORTED: apple-clang-18, but that seem to not have worked. Looking at the [buildkite pipeline configs](https://github.com/llvm/llvm-project/blob/main/libcxx/utils/ci/buildkite-pipeline.yml#L24), it looks like LLVM 18 runs on the apple workers.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (frederick-vs-ja) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:145]: `T[0]` is ill-formed per the standard but widely accepted as an extension. Should we test it? I roughly remember that MSVC STL's configuration for testing make compilers complain about `T[0]`.

Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:145]: Thank you for the comment. I would like to test everything including language extensions (if not too complicated... for example testing VLAs are a bit of a hassle). I'll think what to do (remove or `#ifdef`).

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (Zingam) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:304]: This should be removed: https://github.com/llvm/llvm-project/pull/107238
```suggestion
```

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (philnik777) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:145]: While these tests are nice for clang, I don't think we should test everything here. We don't define how these extensions interact with standard features and we don't claim any level of support for most of this. Quite to the contrary, we've had discussions about supporting some of these in the past and came to the conclusion that we don't want to support them. We're also not adding any additional testing, since these should already be tested by clang. If not, they can go into the clang test suite.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (robincaloudis) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: Nikolas helped me to solve this issue. Basically, `UNSUPPORTED: apple-clang-16, apple-clang-17` does not work because `Apple back-deployment macosx10.{13,15}` and `Apple back-deployment with hardening enabled` both run on AppleClang 15. In addition, there is no AppleClang 17. See [comment](https://github.com/llvm/llvm-project/pull/105946#discussion_r1743613080).

Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: @robincaloudis Thank you! I'll update the PR. In my case the builtin won't be supported even in AppleClang 17 as it is too new.

Comment (philnik777) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:13]: You don't know that, so let's only add it when AppleClang17 actually exists.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp]: @philnik777 Thank you for pointing this out. I removed all test cases testing Language extensions. Please let me know if I misunderstood you and I've gone too far.

**Review by ldionne — APPROVED** [review]
LGTM, please wait for @philnik777 's approval before merging.

**Comment by Zingam** [comment]
> LGTM, please wait for @philnik777 's approval before merging.

Thank you for the review. I think the CI is not using Clang-20 yet with the builtin and I should wait for the CI to be updated before I merge. WDYT?

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (Zingam) [reply libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp]: > You don't know that, so let's only add it when AppleClang17 actually exists.

I mixed the LLVM version with the AppleClang version. Sorry! Thanks!

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (philnik777) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:132]: We've decided to not support `__restrict`, at least without an RFC detailing the implications.

File: libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp
Comment (philnik777) [inline libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp:185]: Here are also a bunch of duplicates. `long`, `long int`, `signed long` and `signed long int` are the same type. Maybe you should just use `arithmetic_types`. That would also remove the need for the `#ifdef`s above.

File: libcxx/docs/ReleaseNotes/20.rst
Comment (Zingam) [inline libcxx/docs/ReleaseNotes/20.rst:46]:
```suggestion
- ``std::jthread`` and ``<stop_token>`` are not guarded behind ``-fexperimental-library`` anymore
- P2674R1: A trait for implicit lifetime types (`Github <https://github.com/llvm/llvm-project/issues/105259>`__)
```

File: libcxx/docs/ReleaseNotes/20.rst
Comment (Zingam) [inline libcxx/docs/ReleaseNotes/20.rst:48]:
```suggestion
- P2985R0: A type trait for detecting virtual base classes (`Github <https://github.com/llvm/llvm-project/issues/105432>`__)
- ``std::jthread`` and ``<stop_token>`` are not guarded behind ``-fexperimental-library`` anymore
- P2674R1: A trait for implicit lifetime types (`Github <https://github.com/llvm/llvm-project/issues/105259>`__)

Improvements and New Features
```

File: libcxx/include/module.modulemap
Comment (Zingam) [inline libcxx/include/module.modulemap:207]: @ldionne The new module map required this PR to be updated. Could you please reapprove the above change?

**Comment by Zingam** [comment]
@philnik777 I addressed your comments. I checked `std::atomic` as you asked and it passes the trait in runtime context. I'm not sure if we need such tests so I didn't include any.

File: libcxx/include/module.modulemap
Comment (philnik777) [inline libcxx/include/module.modulemap:205]: You don't need to export the module itself. Did you mean `std_core.type_traits.integral_constant`?

**Review by philnik777 — COMMENTED** [review]
Basically LGTM except for the modulemap.

File: libcxx/include/module.modulemap
Comment (Zingam) [reply libcxx/include/module.modulemap]: Thanks!

**Comment by Zingam** [comment]
I guess this PR will have to be updated with the changes from https://github.com/llvm/llvm-project/pull/111522

**Review by philnik777 — APPROVED** [review]

**Comment by ldionne** [comment]
Modulemap changes look good to me.

I looked at the sanitizer buildbot thing above and I can't find anything in the logs.

**Pull request merged**
